test/gtest/fold.hpp

```diff
@@ -0,0 +1,306 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTN OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTN WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include "../driver/tensor_driver.hpp"
+#include "cpu_fold.hpp"
+#include "get_handle.hpp"
+#include "miopen/allocator.hpp"
+#include "random.hpp"
+#include "tensor_holder.hpp"
+#include "verify.hpp"
+#include <cstddef>
+#include <cstdlib>
+#include <random>
+#include <gtest/gtest.h>
+#include <miopen/miopen.h>
+#include <miopen/fold.hpp>
+
+struct FoldTestCase
+{
+    size_t N;
+    size_t C;
+    size_t D;
+    size_t H;
+    size_t W;
+    std::vector<int32_t> outputSize;
+    std::vector<int32_t> kernelSize;
+    std::vector<int32_t> stride;
+    std::vector<int32_t> padding;
+    std::vector<int32_t> dilation;
+    bool isContiguous = true;
+    friend std::ostream& operator<<(std::ostream& os, const FoldTestCase& tc)
+    {
+        os << "N:" << tc.N << " C:" << tc.C << " D:" << tc.D << " H:" << tc.H << " W:" << tc.W;
+        os << " output_size:";
+        for(const auto& outs : tc.outputSize)
+            os << outs << " ";
+        os << " kernel_size:";
+        for(const auto& ks : tc.kernelSize)
+            os << ks << " ";
+        os << "stride:";
+        for(const auto& s : tc.stride)
+            os << s << " ";
+        os << "padding:";
+        for(const auto& p : tc.padding)
+            os << p << " ";
+        os << "dilation:";
+        for(const auto& d : tc.dilation)
+            os << d << " ";
+        os << "isContiguous:" << std::boolalpha << tc.isContiguous;
+        return os;
+    }
+
+    std::vector<size_t> GetInput()
+    {
+        if((N != 0) && (C != 0) && (D != 0) && (H != 0) && (W != 0))
+        {
+            return std::vector<size_t>({N, C, D, H, W});
+        }
+        else if((N != 0) && (C != 0) && (H != 0) && (W != 0))
+        {
+            return std::vector<size_t>({N, C, H, W});
+        }
+        else if((N != 0) && (C != 0) && (W != 0))
+        {
+            return std::vector<size_t>({N, C, W});
+        }
+        else if((N != 0) && (W != 0))
+        {
+            return std::vector<size_t>({N, W});
+        }
+        else if((N != 0))
+        {
+            return std::vector<size_t>({N});
+        }
+        else
+        {
+            std::cout << "Error Input Tensor Lengths\n" << std::endl;
+            return std::vector<size_t>({0});
+        }
+    }
```

Comment:
The GetInput() func is quite verbose. I suggest using a vector to store dim instead of 5 fields N, C, D, H, W.

Reply:
resolve in [e76095e](https://github.com/moreh-dev/MIOpen/pull/37/commits/e76095e8fc1771a82978af13018b538b396877d1)